### PR TITLE
Added branch prediction macros and marked allocation errors as unlikely

### PR DESCRIPTION
**Pull request by thedrow**
There are probably a lot more places where these macros should be used but this is a good start.

### CODE REVIEW

**Comment by thedrow** [comment]
@lutter This is ready for review :)

File: src/internal.h [hunk 1]
```diff
@@ -138,6 +138,26 @@
...
+*
+* AUGEAS_UNLIKELY:
+*
+* Macro to flag a code branch as an unlikely branch
+*/
+#ifndef __has_builtin
+#   define __has_builtin(x) (0)
+#endif
+#if defined(__builtin_expect) || __has_builtin(__builtin_expect)
+#   define AUGEAS_LIKELY(x)        (__builtin_expect(!!(x), 1))
```
Comment (lutter) [inline src/internal.h]: Could we just call them `LIKELY` and `UNLIKELY` or is that going to clash with some other include ?

**Comment by lutter** [comment]
I like the idea .. out of curiosity, has that made a perf difference for you ?

File: src/internal.h (hunk 1 above)
Comment (thedrow) [reply src/internal.h]: These macros are always prefixed by the project name since other projects may also define their own set of builtin_expect macros.

**Comment by thedrow** [comment]
I don't see any measurable performance improvement but than again I haven't done anything other than running the dump example.
In any case, this is a good idea to tell the CPU that error branches are rarely executed.

**Comment by lutter** [comment]
Merged as f98b04cc

**Pull request closed without merging**